pkg/virtual/workspaces/registry/rest.go

```diff
@@ -549,3 +570,124 @@ func (s *REST) Delete(ctx context.Context, name string, deleteValidation rest.Va
 }
+
+type KubeconfigSubresourceREST struct {
+	mainRest *REST
+
+	// coreClient is useful to get secrets
+	coreClient corev1client.CoreV1Interface
+	// workspaceShardClient can get KCP workspace shards
+	workspaceShardClient tenancyclient.WorkspaceShardInterface
+}
+
+var _ rest.Getter = &KubeconfigSubresourceREST{}
+var _ rest.Scoper = &KubeconfigSubresourceREST{}
+
+// Get retrieves a Workspace KubeConfig by workspace name
+func (s *KubeconfigSubresourceREST) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	workspace, err := s.mainRest.Get(ctx, name, options)
+	if err != nil {
+		return nil, err
+	}
+	wrapError := func(err error) error {
+		k8sErr := kerrors.NewNotFound(tenancyv1alpha1.SchemeGroupVersion.WithResource("workspaces/kubeconfig").GroupResource(), name)
+		k8sErr.Status().Details.Causes = append(k8sErr.Status().Details.Causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeUnexpectedServerResponse,
+			Message: err.Error(),
+		})
+		return k8sErr
+	}
+	scope := ctx.Value(WorkspacesScopeKey).(string)
+	if workspace, isAWorkspace := workspace.(*tenancyv1alpha1.Workspace); isAWorkspace {
+		if conditions.IsTrue(workspace, tenancyv1alpha1.WorkspaceURLValid) {
+			ctx := genericapirequest.WithCluster(context.TODO(), genericapirequest.Cluster{Name: workspace.ClusterName})
+			shard, err := s.workspaceShardClient.Get(ctx, workspace.Status.Location.Current, metav1.GetOptions{})
+			if err != nil {
+				return nil, wrapError(err)
+			}
+			secret, err := s.coreClient.Secrets(shard.Spec.Credentials.Namespace).Get(ctx, shard.Spec.Credentials.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, wrapError(err)
+			}
+			data, ok := secret.Data[tenancyv1alpha1.WorkspaceShardCredentialsKey]
+			if !ok {
+				return nil, wrapError(fmt.Errorf("Key '%s' not found in workspace shard Kubeconfig secret", tenancyv1alpha1.WorkspaceShardCredentialsKey))
+			}
+			shardKubeConfig, err := clientcmd.Load(data)
+			if err != nil {
+				return nil, wrapError(fmt.Errorf("Workspace shard Kubeconfig is invalid: %w", err))
+			}
+
+			currentContext := shardKubeConfig.Contexts[shardKubeConfig.CurrentContext]
+			if currentContext == nil {
+				return nil, wrapError(errors.New("Workspace shard Kubeconfig has no current context"))
+			}
+			currentCluster := shardKubeConfig.Clusters[currentContext.Cluster]
+			if currentCluster == nil {
+				return nil, wrapError(fmt.Errorf("Workspace shard Kubeconfig has no cluster corresponding to the current context cluster key: %s", currentContext.Cluster))
+			}
+			currentCluster.Server = workspace.Status.BaseURL
+
+			workspaceContextName := scope + "/" + workspace.Name
+			workspaceConfig := &api.Config{
+				APIVersion:     "v1",
+				Clusters:       map[string]*api.Cluster{workspaceContextName: currentCluster},
+				Contexts:       map[string]*api.Context{workspaceContextName: {Cluster: workspaceContextName}},
+				CurrentContext: workspaceContextName,
+			}
```

Comment:
please add some go doc:

// return kubeconfig that lacks the user and its credentials, i.e. it's only the cluster definition
// with its CA cert and URL.

Reply:
Fixed in commit 2c97666d51d66ccc901f7d6f92992f5ba1412a5d